SecurityRule can block URL as 403 forbidden

akollegger opened the pull request:
This adds `SecurityRule.isForbidden()` to allow a rule to block access to a URL regardless of their authentication status, resulting in a `403 forbidden` response.

thobe commented on community/server/src/main/java/org/neo4j/server/rest/security/SecurityRule.java:
```diff
@@ -30,11 +30,18 @@
+     * @return <code>true</code> if the rule is triggered and the request is to be rejected with a "403 Forbidden",
+     *  and <code>false</code> otherwise.
+     */
+    boolean isForbidden(HttpServletRequest request);
```
`org.neo4j.server.rest.security.SecurityRule` is an interface that users are [expected to implement](http://neo4j.com/docs/stable/security-server.html#_server_authorization_rules), adding a new method constitutes an incompatible change.

I would say that we cannot make this change until 3.0, for two reasons:
1. We allow ourselves to make such breaking changes on major releases.
2. Most likely we will be using Java 8 for 3.0 at which point we could provide a default implementation for this method.

akollegger replied:
Agreed, a breaking change is not acceptable. Revising.

thobe commented on packaging/neo4j-desktop/neo4j-home/ssl/snakeoil.cert:
```diff
@@ -0,0 +1,10 @@
+SIb3DQEBBQUAA4GBAB4ztTTPvJeHcDWr9xEWh+tYbnB78iD56CQjgQJpfikjC9HXJJfjze+r9hlp
+SlSj8q9URk7ysDiWStqanxcWz1gAsEa+oVFxLOjMY5/A6O+HcqNR36FBDTphYZJUe0peTgwynnmB
+ERBfFu3JQ8evBg8w474OTlbGbag6O0xjFV5x
+-----END CERTIFICATE-----
```
Did you really intend to check in this file?

akollegger replied:
Um, no. Looks like a side-effect of building. I'll add to `.gitignore`

thobe commented on packaging/neo4j-desktop/neo4j-home/ssl/snakeoil.key:
```diff
@@ -0,0 +1,14 @@
+G0Dy0M+cX5PT50K9aAh6uj+SrqWdEf2FmJQefCRMIXq3YSX/mJ4MKytydebkMGPhwaP/RFTNAkEA
+sBR2B5Le4A76Lozrunm2EJb2oqTuk02s9SWVx0Q4YN/uWIwjmp1YkChcTbMwSM2MwoPpDhjBRXPb
+wa/xAdjxpQ==
+-----END PRIVATE KEY-----
```
Did you really intend to check in this file?

davidegrohmann commented:
@akollegger @thobe what's the status of this PR?

thobe commented:
@davidegrohmann it is not acceptable in its current form, since it breaks backwards compatibility. @akollegger has said that he will revise it, but if that will not happen we should close this PR.

akollegger committed:
Introduced ForbiddingSecurityRule

rather than change the contract of SecurityRule, a facade was introduced to apply any SecurityRule variants which may be implemented. The first variant is ForbiddingSecurityRule, which can result in a 403 for a request.
